design-docs/improved-console.md

```diff
@@ -0,0 +1,283 @@
+
+Plugins can provide progress indication by publishing `BuildOperations`
+
+A `OperationStatusBarFormatter` can be used to customize the intra-line display of each 
+`ProgressOperation` chain (each `ProgressOperation` has a reference to its parent)
+
+Implementation is similar to `ConsoleBackedProgressRenderer` with throttling and a data
+structure for storing recent updates.
+
+> NOTE: We must avoid use of save/restore cursor position unless we switch to a 
+[terminfo](https://en.wikipedia.org/wiki/Terminfo)-based Console handling instead of using ANSI.
+    
+### Test Coverage
+* A terminal size with fewer than max parallel operations shows (rows - 2) operations in parallel
+* Operation status lines are trimmed at (cols - 1)
+* `System.in` and SystemConsole I/O happens on the mainArea, which is unaffected by other areas
+* Renders well on default macOS Terminal, default Windows Console, Cygwin, and common Linux Terminals
+* Console display degrades gracefully given BuildOperations that are out-of-order (e.g. complete event received before start)
+
+### Open issues
+* How is `--continuous` build displayed? During execution? While waiting? (as we allow processes to stay running)
+* When things fail, will the new CLI look like a regular failure or something else?
+* How does `--parallel` (or not) affect display?
+* How does having more operations than workers affect display?
+
+## Story: Display build progress as a progress bar through colorized output
+Render ProgressEvents with build completeness information as a progress bar.
+ 
+Intended to give the user a very fast way of telling whether the build will be finished soon or not.
+The work-in-progress (yellow) section shows how much of the build would be complete if all
+Operations displayed in the progressArea completed.
+
+### User-visible Changes
+We can visually represent complete, in-progress, and un-started tasks of build using colorized output:
+
+`#####green#####>##yellow##>#####black#####> 40% Building`
+
+On ANSI terminals, we can use empty spaces with different background colors for the "bars" and
+A red background shall be used for failed tasks. 
+
+### Implementation
+`DefaultGradleLauncherFactory` registers a `BuildProgressListener` listening
+to events from `BuildProgressLogger` and forwards them to a `ConsoleBackedBuildProgressRenderer`.
+ 
+`ProgressLine` consists of a `ProgressBar`, status text, and a width. It doesn't know anything about builds. 
+
+`ProgressBar` has text and width. Weakly modeling this allows for alternate `ProgressLine`s that can
+potentially be configured by users.
+
+The `DefaultProgressBar`, though, has a width and a list of `ProgressBarRegion`s,
+each of which have a width and a `RegionType` of `SUCCEEDED`, `FAILED`, `IN_PROGRESS`, or `UNSTARTED`. 
+Each`RegionType` is associated with a color. This allows us to show failures in arbitrary locations
+as needed with `--continue`
+
+### Test Coverage
```

Comment:
Neither CI or IDEs attach to a Console, so their behavior will not change at all except for additional information they'll see from increased BuildOperations.